### ISSUE TITLE
pgch is not threadsafe

### ISSUE THREAD

**Issue by ahelsing** [GENI-NSF/geni-portal#690]
Using Omni, Luisa got a weird error where she couldn't get her slice credential cause pgch thought her user credential was for lnevers but her certificate was for jbs?

```
13:37:57 ERROR    omni.pgch: Call for Get Slice Cred for slice urn:publicid:IDN+ch.geni.net:ln-prj+slice+lnstitch failed.: Exception: Failed to get GENI Clearinghouse slice urn:publicid:IDN+ch.geni.net:ln-prj+slice+lnstitch credential: Error: 1, Message: <Fault Insufficient privileges: '<Fault Insufficient privileges: "No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+lnevers. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+lnevers fails: Credential doesn\'t grant rights to you (urn:publicid:IDN+ch.geni.net+user+jbs), but to urn:publicid:IDN+ch.geni.net+user+lnevers (over object urn:publicid:IDN+ch.geni.net+user+lnevers)">'>
```

 _Imported from trac ticket [#690](http://trac.gpolab.bbn.com/proto-ch/ticket/690),  created by ahelsing on 07-03-2013 at 15:54, last modified: 07-09-2013 at 10:18_

**Comment by ahelsing** [comment]
```
[Wed Jul 03 13:37:39 2013] [INFO] gcf-pgch: Found slice id 63355e13-c843-4152-89a7-9b8e086c803f for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps101
[Wed Jul 03 13:37:40 2013] [INFO] gcf-pgch: Found slice id 63355e13-c843-4152-89a7-9b8e086c803f for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps101
[Wed Jul 03 13:37:45 2013] [INFO] gcf-pgch: Found slice id 63355e13-c843-4152-89a7-9b8e086c803f for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps101
[Wed Jul 03 13:37:51 2013] [INFO] gcf-pgch: Found slice id 1f4bb094-290e-4443-9888-7aff62db62e3 for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps102
[Wed Jul 03 13:37:53 2013] [INFO] gcf-pgch: Found slice id 1f4bb094-290e-4443-9888-7aff62db62e3 for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps102
[Wed Jul 03 13:37:57 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+lnevers. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+lnevers fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+jbs), but to urn:publicid:IDN+ch.geni.net+user+lnevers (over object urn:publicid:IDN+ch.geni.net+user+lnevers)
[Wed Jul 03 13:37:57 2013] [INFO] gcf-pgch: Found slice id 1f4bb094-290e-4443-9888-7aff62db62e3 for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps102
[Wed Jul 03 13:37:58 2013] [INFO] gcf-pgch: Found slice id 1f4bb094-290e-4443-9888-7aff62db62e3 for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps102
[Wed Jul 03 13:38:00 2013] [INFO] gcf-pgch: Found slice id 1f4bb094-290e-4443-9888-7aff62db62e3 for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps102
[Wed Jul 03 13:38:02 2013] [INFO] gcf-pgch: Found slice id 1f4bb094-290e-4443-9888-7aff62db62e3 for urn urn:publicid:IDN+ch.geni.net:JBS+slice+ps102
```


_Trac comment by ahelsing on 07-03-2013 at 15:54_

**Comment by ahelsing** [comment]
The logs show many cases of this since right after we installed this new threading code:
```
[Wed Jul 03 12:05:40 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+lnevers. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+lnevers fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+mbrinn), but to urn:publicid:IDN+ch.geni.net+user+lnevers (over object urn:publicid:IDN+ch.geni.net+user+lnevers)
[Wed Jul 03 12:05:47 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+mbrinn. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+mbrinn fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+lnevers), but to urn:publicid:IDN+ch.geni.net+user+mbrinn (over object urn:publicid:IDN+ch.geni.net+user+mbrinn)
[Wed Jul 03 12:23:42 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+jbs. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+jbs fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+charles), but to urn:publicid:IDN+ch.geni.net+user+jbs (over object urn:publicid:IDN+ch.geni.net+user+jbs)
[Wed Jul 03 12:23:43 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+charles. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+charles fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+jbs), but to urn:publicid:IDN+ch.geni.net+user+charles (over object urn:publicid:IDN+ch.geni.net+user+charles)
[Wed Jul 03 12:25:47 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+charles. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+charles fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+jbs), but to urn:publicid:IDN+ch.geni.net+user+charles (over object urn:publicid:IDN+ch.geni.net+user+charles)
[Wed Jul 03 13:19:57 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+cwang1. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+cwang1 fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+jbs), but to urn:publicid:IDN+ch.geni.net+user+cwang1 (over object urn:publicid:IDN+ch.geni.net+user+cwang1)
[Wed Jul 03 13:37:09 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+jbs. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+jbs fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+lnevers), but to urn:publicid:IDN+ch.geni.net+user+jbs (over object urn:publicid:IDN+ch.geni.net+user+jbs)
[Wed Jul 03 13:37:57 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+lnevers. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+lnevers fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+jbs), but to urn:publicid:IDN+ch.geni.net+user+lnevers (over object urn:publicid:IDN+ch.geni.net+user+lnevers)
[Wed Jul 03 13:55:47 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+cwang1. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+cwang1 fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+amcanary), but to urn:publicid:IDN+ch.geni.net+user+cwang1 (over object urn:publicid:IDN+ch.geni.net+user+cwang1)
[Wed Jul 03 14:39:42 2013] [ERROR] cred-verifier: No credential was found with appropriate privileges. Tried urn:publicid:IDN+ch.geni.net+user+kobus. Last failure: Cred urn:publicid:IDN+ch.geni.net+user+kobus fails: Credential doesn't grant rights to you (urn:publicid:IDN+ch.geni.net+user+lnevers), but to urn:publicid:IDN+ch.geni.net+user+kobus (over object urn:publicid:IDN+ch.geni.net+user+kobus)
```


_Trac comment by ahelsing on 07-03-2013 at 15:56_

**Comment by ahelsing** [comment]
Is there some way that the SecureXMLRPCRequestHandler's request member (which is not thread safe) is changing while SecureThreadedXMLRPCRequestHandler.setup() is executing? Maybe we need a lock on the setup() call?

Note that this wouldn't have appeared in Marshall's testing, cause it was always the same user cert. Testing this requires using multiple separate accounts at the same time.

_Trac comment by ahelsing on 07-03-2013 at 16:24_

**Comment by ahelsing** [comment]
pgch.py does:
```
./src/geni/pgch.py:        user_certstr = addMACert(self._server.pem_cert, self.logger, self.macert)
```

In the threaded world, that is wrong. I suspect that pgch can do if THREADING get_pem_cert() or some such.

pgch also over-rides the request handler:
```
class PgChRequestHandler(SecureXMLRPCRequestHandler):
```

so we're not getting the thread local at all


_Trac comment by ahelsing on 07-03-2013 at 16:48_

**Comment by ahelsing** [comment]
I disabled threading for now (tag 1.9.2). Note this included installing a local copy of ch.py - we should remove this when we have the threading working correctly.


_Trac comment by ahelsing on 07-03-2013 at 17:14_

**Comment by ahelsing** [comment]
I believe that to be safe the threaded version might also need to:
 - add a line in setup() to unset the server.peercert and server.der_cert and server.pem_cert
 - add a call to finish() that cleans up the thread local?
 - setup(): consider a lock around the 2 lines that each read the cert off of the request, to make certain that the request doesn't somehow change while reading this data

To test this, we should try to run a multithreaded GCF pgch. Then create 10 gcf-ch users, and have them all make pgch calls simultaneously repeatedly.
The fix can go on the gcf/gcf-2.2.1-threading and gcf/develop branches for ch.py, and on proto-ch/develop-gec17 and gcf/develop for pgch.py


_Trac comment by ahelsing on 07-05-2013 at 12:01_

**Comment by ahelsing** [comment]
I reproduced the issue using the gcf backed pgch and a lot of background omni calls to getusercred, having hacked the code a little to show the issue. 
I made Omni (framework_pg) pass along the name of the user cert file in the call to GetCredential (which I conveniently named differently for each user). And I made GetCredential print the user cert arg it got, and the HRN of the user whose cert it pulled from server.pem_cert
This is the printout from pgch
```
GetCred got arg of user cert /home/ahelsing/.gcf/carl-cert.pem, server.pem_cert for user geni.gpo.gcf.jon
```

Having reproduced it, I can try to fix it.


_Trac comment by ahelsing on 07-05-2013 at 16:38_
